fx quant: fix subtle bug in BinaryOpQuantizeHanlder logic in matching

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56004 [quant] added dq->op->q quantization patterns for GELU and softmax ops
* **#56294 fx quant: fix subtle bug in BinaryOpQuantizeHanlder logic in matching**

Summary:

When matching a pattern to `BinaryOpQuantizeHandler`, we need to make
sure we check for dtype support on the base node, instead of the current
node.  This is important in cases such as `add-relu` and `mul-relu`,
when the current node is `relu`, but the base node is `add|mul`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx
```

There is no good test case to check this in current logic.  Created an
add-relu model manually, and verified with pdb that the add node was
being used to match against dtypes.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27831070](https://our.internmc.facebook.com/intern/diff/D27831070)

vkuzo committed:
fx quant: fix subtle bug in BinaryOpQuantizeHanlder logic in matching

Summary:

When matching a pattern to `BinaryOpQuantizeHandler`, we need to make
sure we check for dtype support on the base node, instead of the current
node.  This is important in cases such as `add-relu` and `mul-relu`,
when the current node is `relu`, but the base node is `add|mul`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx
```

There is no good test case to check this in current logic.  Created an
add-relu model manually, and verified with pdb that the add node was
being used to match against dtypes.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "fx quant: fix subtle bug in BinaryOpQuantizeHanlder logic in matching"


Summary:

When matching a pattern to `BinaryOpQuantizeHandler`, we need to make
sure we check for dtype support on the base node, instead of the current
node.  This is important in cases such as `add-relu` and `mul-relu`,
when the current node is `relu`, but the base node is `add|mul`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx
```

There is no good test case to check this in current logic.  Created an
add-relu model manually, and verified with pdb that the add node was
being used to match against dtypes.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27831070](https://our.internmc.facebook.com/intern/diff/D27831070)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -1313,7 +1313,20 @@ def record_match(pattern, node, matched):
                                 value = CopyNodeQuantizeHandler  # type: ignore
 
-                            this_node_qconfig = self.qconfig_map[node.name]
+                            # to properly check for dtype support, we need to
```
is this the case when base_node is configured to be quantized and relu is not?

vkuzo replied:
this is because dtypes supported by relu can be different than dtypes supported by add/mul/sub/etc, and since we are in `BinaryOpQuantizeHandler`, the dtypes to use need to be the ones of the base node.

vkuzo committed:
Update on "fx quant: fix subtle bug in BinaryOpQuantizeHanlder logic in matching"


Summary:

When matching a pattern to `BinaryOpQuantizeHandler`, we need to make
sure we check for dtype support on the base node, instead of the current
node.  This is important in cases such as `add-relu` and `mul-relu`,
when the current node is `relu`, but the base node is `add|mul`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx
```

There is no good test case to check this in current logic.  Created an
add-relu model manually, and verified with pdb that the add node was
being used to match against dtypes.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27831070](https://our.internmc.facebook.com/intern/diff/D27831070)

[ghstack-poisoned]

HDCharles committed:
Update on "fx quant: fix subtle bug in BinaryOpQuantizeHanlder logic in matching"


Summary:

When matching a pattern to `BinaryOpQuantizeHandler`, we need to make
sure we check for dtype support on the base node, instead of the current
node.  This is important in cases such as `add-relu` and `mul-relu`,
when the current node is `relu`, but the base node is `add|mul`.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx
```

There is no good test case to check this in current logic.  Created an
add-relu model manually, and verified with pdb that the add node was
being used to match against dtypes.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27831070](https://our.internmc.facebook.com/intern/diff/D27831070)

[ghstack-poisoned]
